### PR TITLE
Add logout prehooks

### PR DESCRIPTION
**Pull request by sharidas**
This change will help to do actions
in logout prehooks. Any changes to
be made before logout action can be
done here.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

Add logout prehooks

## Motivation and Context

Adding logout prehooks

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: lib/private/Server.php [hunk 1]
```diff
@@ -322,6 +323,10 @@ public function __construct($webRoot, \OC\Config $config) {
 				/** @var $user \OC\User\User */
 				\OC_Hook::emit('OC_User', 'post_login', ['run' => true, 'uid' => $user->getUID(), 'password' => $password]);
 			});
+			$userSession->listen('\OC\User', 'preLogout', function () {
+				$event = new GenericEvent(null, []);
+				\OC::$server->getEventDispatcher()->dispatch('prelogoutHook', $event);
```
Comment (karakayasemi) [inline lib/private/Server.php]: optional, but we can consider following the naming convention of EventDispatcher.

File: lib/private/User/Session.php [hunk 2]
```diff
@@ -820,10 +820,27 @@ public function loginWithCookie($uid, $currentToken) {
...
+
+	/**
 	 * logout the user from the session
 	 */
 	public function logout() {
-		$this->manager->emit('\OC\User', 'logout');
+		if ($this->preLogout() === true) {
+			return;
+		}
+			$this->manager->emit('\OC\User', 'logout');
```
Comment (karakayasemi) [inline lib/private/User/Session.php]: indentation

File: lib/private/Server.php (hunk 1 above)
Comment (PVince81) [reply lib/private/Server.php]: please do

File: lib/private/User/Session.php [hunk 3]
```diff
@@ -823,6 +823,13 @@ public function loginWithCookie($uid, $currentToken) {
 	 * logout the user from the session
 	 */
 	public function logout() {
+		$this->manager->emit('\OC\User', 'preLogout');
+
+		if ($this->getSession()->get('switchedToNewUser') === 'yes') {
```
Comment (PVince81) [inline lib/private/User/Session.php]: Don't abuse the session to pass around variables. There is an existing way to receive values from a symfony event, just use it. Here's an example: https://github.com/owncloud/guests/blob/master/controller/userscontroller.php#L130

**Comment by sharidas** [comment]
@PVince81  No more sessions. Data handled via Symfony events.

File: lib/private/User/Session.php [hunk 4]
```diff
@@ -823,6 +823,21 @@ public function loginWithCookie($uid, $currentToken) {
 	 * logout the user from the session
 	 */
 	public function logout() {
+		$called = [];
+
+		$eventDispatcher = \OC::$server->getEventDispatcher();
+		$eventDispatcher->addListener('OC\User\Session::logout', function ($event) use (&$called) {
+			if ($event['switchedToNewUser'] === 'true') {
```
Comment (PVince81) [inline lib/private/User/Session.php]: I think this should be called "cancel".

Also why not use booleans instead of strings containing booleans ?

**Comment by sharidas** [comment]
@PVince81 Updated the new patch by making 'cancel' and used booleans instead of string containing booleans. Thanks for the feedback.

File: lib/private/User/Session.php [hunk 5]
```diff
@@ -823,6 +823,19 @@ public function loginWithCookie($uid, $currentToken) {
 	 * logout the user from the session
 	 */
 	public function logout() {
+		$cancelLogout = false;
+
+		$eventDispatcher = \OC::$server->getEventDispatcher();
+		$eventDispatcher->addListener('OC\User\Session::logout', function ($event) use (&$cancelLogout) {
+			$cancelLogout = $event['cancel'];
```
Comment (PVince81) [inline lib/private/User/Session.php]: Have you tested this ? You are not even firing the event with `$eventDispatcher`

Please look again at the example I posted and use the same approach

Comment (sharidas) [reply lib/private/User/Session.php]: Its been fired from https://github.com/owncloud/impersonate/pull/68/files#diff-5ebb5b2aa16f3a5e11cb19da41fe2f7dR77.

Comment (PVince81) [reply lib/private/User/Session.php]: Ah, that's a misunderstanding. Core must fire the event and apps must listen to it to react, not the other way around.

**Comment by sharidas** [comment]
@PVince81 Updated the change to trigger the event from core. And the impersonate app reacts to it. Thanks for the feedback.

**Comment by sharidas** [comment]
@PVince81 I have used naming convention for the event names. Let me know if this change is ok.

File: lib/private/Server.php [hunk 6]
```diff
@@ -322,6 +323,10 @@ public function __construct($webRoot, \OC\Config $config) {
 				/** @var $user \OC\User\User */
 				\OC_Hook::emit('OC_User', 'post_login', ['run' => true, 'uid' => $user->getUID(), 'password' => $password]);
 			});
+			$userSession->listen('\OC\User', 'preLogout', function () {
+				$event = new GenericEvent(null, []);
+				\OC::$server->getEventDispatcher()->dispatch('pre_logoutHook', $event);
```
Comment (karakayasemi) [inline lib/private/Server.php]: "user.pre_logout_hook" is better i guess. According to the naming convention:
-Use only lowercase letters, numbers, dots (.) and underscores (_);
-Prefix names with a namespace followed by a dot (e.g. order., user.*);
-End names with a verb that indicates what action has been taken (e.g. order.placed).

Comment (sharidas) [reply lib/private/Server.php]: Done

File: lib/private/Server.php [hunk 7]
```diff
@@ -322,6 +323,10 @@ public function __construct($webRoot, \OC\Config $config) {
 				/** @var $user \OC\User\User */
 				\OC_Hook::emit('OC_User', 'post_login', ['run' => true, 'uid' => $user->getUID(), 'password' => $password]);
 			});
+			$userSession->listen('\OC\User', 'preLogout', function () {
+				$event = new GenericEvent(null, []);
+				\OC::$server->getEventDispatcher()->dispatch('user.pre_logout_hook', $event);
```
Comment (PVince81) [inline lib/private/Server.php]: hmmm, is this the official Symfony naming convention ? for all other Symfony events in OC we used a namespaced name, please look it up

**Review by PVince81 — COMMENTED** [review]
please double check namespace format and add unit tests to make codecov happy.

File: lib/private/Server.php (hunk 7 above)
Comment (karakayasemi) [reply lib/private/Server.php]: https://symfony.com/doc/current/components/event_dispatcher.html#naming-conventions
Yes, this is the official Symfony naming convention, but we don't have to follow that of course. I checked quickly the other Symfony events in OC. Generally, we used a namespaced name, but 3rdparty libraries are following this convention, like:
https://github.com/owncloud/core/blob/5ce66c4d5f9a895bec41f56f7cc96661eec00ae6/apps/files_external/3rdparty/aws-sdk-php/Aws/Common/Model/MultipartUpload/AbstractTransfer.php#L33
or
https://github.com/owncloud/core/blob/5ce66c4d5f9a895bec41f56f7cc96661eec00ae6/apps/files_external/3rdparty/aws-sdk-php/Guzzle/Plugin/Backoff/BackoffPlugin.php#L19
All the events related to a class should be constant class property and we should use these constants when dispatching an event. IMHO it is cleaner. 
By the way, core have some codes that follow the convention:
https://github.com/owncloud/core/blob/baf07a7755009e76dcb10cf243330070ec41f194/apps/files_sharing/lib/API/Remote.php#L116
or
https://github.com/owncloud/core/blob/master/lib/public/Http/HttpEvents.php#L32

Comment (sharidas) [reply lib/private/Server.php]: For code consistency I am modifying the event name to use namespace.

Comment (PVince81) [reply lib/private/Server.php]: @DeepDiver1975 any comment on this ? I assume we want to continue using the namespace format like we did for loadAdditionalScripts

Comment (karakayasemi) [reply lib/private/Server.php]: AFAIK, we are still in the adaptation phase to EventDispatcher. IMHO, we can add event strings as constant property to the related class, we can track it under #29174 and follow the convention.

File: lib/private/User/Session.php [hunk 8]
```diff
@@ -821,8 +822,21 @@ public function loginWithCookie($uid, $currentToken) {
...
 	/**
 	 * logout the user from the session
+	 *
+	 * @return bool
 	 */
 	public function logout() {
+
+		$event = new GenericEvent(null, ['cancel' => false]);
+		$eventDispatcher = \OC::$server->getEventDispatcher();
+		$eventDispatcher->dispatch('\OC\User\Session::cancel_logout', $event);
```
Comment (PVince81) [inline lib/private/User/Session.php]: No, the event should be `\OC\User\Session::pre_logout`. We are informing every listener that we are about to logout, not about the cancel.

File: lib/private/Server.php [hunk 9]
```diff
@@ -322,6 +323,10 @@ public function __construct($webRoot, \OC\Config $config) {
 				/** @var $user \OC\User\User */
 				\OC_Hook::emit('OC_User', 'post_login', ['run' => true, 'uid' => $user->getUID(), 'password' => $password]);
 			});
+			$userSession->listen('\OC\User', 'preLogout', function () {
+				$event = new GenericEvent(null, []);
+				\OC::$server->getEventDispatcher()->dispatch('\OC\User\Session::pre_logouthook', $event);
```
Comment (PVince81) [inline lib/private/Server.php]: "pre_logout", remove the "hook" from the name

**Comment by sharidas** [comment]
@PVince81 Hope the code looks better.

**Review by PVince81 — APPROVED** [review]
:+1: thanks, please backport

**Comment by sharidas** [comment]
Backport PR: https://github.com/owncloud/core/pull/29352

**Pull request merged**